File: ios/DoMemory/DoMemory/Modules/SharedModules/Views/LivesRow.swift
//
//  LivesRow.swift
//  DoMemory
//
//  Row of heart glyphs showing remaining Levels-mode lives. Reused by the
//  Levels map header and the out-of-lives prompts (LoseModal, OutOfLivesModal).
//

import SwiftUI

/// A one-shot effect the row plays over a single heart: the heart that was
/// just lost breaking apart, or the one just refilled bursting back in.
enum LivesRowEffect: Equatable {
    case lost(slot: Int)
    case gained(slot: Int)

    /// The effect a change in the remaining count deserves, if any. A drop
    /// breaks the first now-empty heart; a rise bursts the last now-filled
    /// one. A day-reset refill from 0 to 4 animates a single heart rather
    /// than four — one burst is a nicety, four is a fireworks show.
    static func forTransition(from previous: Int, to current: Int) -> LivesRowEffect? {
        if current < previous { return .lost(slot: current) }
        if current > previous { return .gained(slot: current - 1) }
        return nil
    }
}

struct LivesRow: View {
    var remaining: Int
    var total: Int = LevelLivesService.maxLives
    var iconSize: CGFloat = 16
    /// Plays once over the named slot. The row draws its real final state
    /// underneath — the clips end fully transparent — so a reduce-motion
    /// player, who never sees the clip, sees exactly the same hearts.
    var effect: LivesRowEffect? = nil
    var onEffectFinished: (() -> Void)? = nil
    /// A still-filled heart that a pending loss will take unless the player
    /// rescues the game. It pulses rather than breaking, because it hasn't
    /// gone yet.
    var atRiskSlot: Int? = nil

    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    /// The heart briefly scaled up by a refill, so the burst has a heart
    /// physically popping back in at its centre rather than just decorating one.
    @State private var bumpedSlot: Int? = nil

    var body: some View {
        HStack(spacing: 4) {
            ForEach(0..<total, id: \.self) { index in
                heart(at: index)
                    .scaleEffect(bumpedSlot == index ? 1.35 : 1)
                    .animation(.spring(response: 0.3, dampingFraction: 0.5), value: bumpedSlot)
                    .overlay { overlay(for: index) }
            }
        }
        .accessibilityElement()
        .accessibilityLabel(Strings.livesRemainingFormat(remaining, total))
    }

    @ViewBuilder
    private func heart(at index: Int) -> some View {
        let glyph = Image(systemName: index < remaining ? "heart.fill" : "heart")
            .font(.system(size: iconSize, weight: .semibold))
            .foregroundStyle(index < remaining ? Color.secundaryColor : Color.textMuted.opacity(0.3))

        if index == atRiskSlot, index < remaining {
            if reduceMotion {
                glyph.opacity(0.45)
            } else {
                glyph.phaseAnimator([false, true]) { content, dimmed in
                    content
                        .opacity(dimmed ? 0.3 : 1)
                        .scaleEffect(dimmed ? 0.85 : 1)
                } animation: { _ in
                    .easeInOut(duration: 0.7)
                }
            }
        } else {
            glyph
        }
    }

    @ViewBuilder
    private func overlay(for index: Int) -> some View {
        if !reduceMotion, let effect {
            switch effect {
            case .lost(let slot) where slot == index:
                // The clip's heart spans 60% of its canvas; 1.7x the glyph
                // size lands it on top of the SF Symbol before it splits.
                LottieView(name: "heart-break", tint: Color.secundaryColor) { onEffectFinished?() }
                    .frame(width: iconSize * 1.7, height: iconSize * 1.7)
                    .allowsHitTesting(false)
            case .gained(let slot) where slot == index:
                LottieView(name: "heart-refill", tint: Color.secundaryColor) { onEffectFinished?() }
                    .frame(width: iconSize * 3, height: iconSize * 3)
                    .allowsHitTesting(false)
                    .task {
                        bumpedSlot = index
                        try? await Task.sleep(nanoseconds: 250_000_000)
                        bumpedSlot = nil
                    }
            default:
                EmptyView()
            }
        }
    }
}

#Preview {
    VStack(spacing: 16) {
        LivesRow(remaining: 4)
        LivesRow(remaining: 2)
        LivesRow(remaining: 0)
        LivesRow(remaining: 2, effect: .lost(slot: 2))
        LivesRow(remaining: 3, effect: .gained(slot: 2))
        LivesRow(remaining: 3, atRiskSlot: 2)
    }
    .padding()
}

File: android/app/src/main/java/com/ezequielbrrt/domemory/ui/components/LivesRow.kt
package com.ezequielbrrt.domemory.ui.components

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.sp
import com.ezequielbrrt.domemory.R
import com.ezequielbrrt.domemory.feature.levels.LivesEffect
import com.ezequielbrrt.domemory.services.levels.LevelLivesService
import com.ezequielbrrt.domemory.ui.anim.rememberReduceMotion
import com.ezequielbrrt.domemory.ui.lottie.BundledLottie
import com.ezequielbrrt.domemory.ui.theme.LocalPalette
import kotlinx.coroutines.delay

/**
 * Matches iOS's `LivesRow`: a compact row of filled and outline heart glyphs, with an
 * optional one-shot [effect] played over a single heart.
 *
 * The row always draws its real final state underneath — the clips end fully
 * transparent — so a reduce-motion player, who never sees the clip, sees exactly the same
 * hearts. [onEffectFinished] is the caller's cue to clear [effect] so it does not replay
 * on the next recomposition.
 */
@Composable
fun LivesRow(
    remaining: Int,
    modifier: Modifier = Modifier,
    total: Int = LevelLivesService.MAX_LIVES,
    fontSize: TextUnit = 16.sp,
    effect: LivesEffect? = null,
    onEffectFinished: (() -> Unit)? = null,
    /** A still-filled heart a pending loss will take unless the player rescues the game.
     * It pulses rather than breaking, because it hasn't gone yet (iOS's `atRiskSlot`). */
    atRiskSlot: Int? = null,
) {
    val palette = LocalPalette.current
    val reduceMotion = rememberReduceMotion()
    val accessibilityLabel = stringResource(R.string.levels_lives_remaining_format, remaining, total)
    // The heart briefly scaled up by a refill, so the burst has a heart physically popping
    // back in at its centre rather than just decorating one.
    var bumpedSlot by remember { mutableStateOf<Int?>(null) }
    val activeEffect = effect?.takeUnless { reduceMotion }
    val atRiskAlpha = if (atRiskSlot == null) {
        1f
    } else if (reduceMotion) {
        0.45f
    } else {
        val pulse by rememberInfiniteTransition(label = "atRiskHeart").animateFloat(
            initialValue = 1f,
            targetValue = 0.3f,
            animationSpec = infiniteRepeatable(tween(700), RepeatMode.Reverse),
            label = "atRiskHeartAlpha",
        )
        pulse
    }

    LaunchedEffect(activeEffect) {
        if (activeEffect is LivesEffect.Gained) {
            bumpedSlot = activeEffect.slot
            delay(250)
            bumpedSlot = null
        }
    }

    Row(
        horizontalArrangement = Arrangement.spacedBy(4.dp),
        modifier = modifier.clearAndSetSemantics { contentDescription = accessibilityLabel },
    ) {
        repeat(total) { index ->
            val bump by animateFloatAsState(
                targetValue = if (bumpedSlot == index) 1.35f else 1f,
                animationSpec = spring(dampingRatio = 0.5f, stiffness = 440f),
                label = "lifeBump",
            )
            Box(contentAlignment = Alignment.Center) {
                Text(
                    text = if (index < remaining) "♥" else "♡",
                    fontSize = fontSize,
                    fontWeight = FontWeight.SemiBold,
                    color = if (index < remaining) palette.secondary else palette.textSecondary.copy(alpha = 0.3f),
                    modifier = Modifier.graphicsLayer {
                        val atRisk = index == atRiskSlot && index < remaining
                        // Dims and shrinks together, like iOS's phase-animated heart.
                        val pulse = if (atRisk) atRiskAlpha else 1f
                        alpha = pulse
                        scaleX = bump * (0.85f + 0.15f * pulse)
                        scaleY = bump * (0.85f + 0.15f * pulse)
                    },
                )
                if (activeEffect != null && activeEffect.slot == index) {
                    val glyph = with(androidx.compose.ui.platform.LocalDensity.current) { fontSize.toDp() }
                    when (activeEffect) {
                        // The clip's heart spans 60% of its canvas; 1.7x the glyph size
                        // lands it on top of the text heart before it splits.
                        is LivesEffect.Lost -> BundledLottie(
                            name = "heart-break",
                            tint = palette.secondary,
                            modifier = Modifier.size(glyph * 1.7f),
                            onFinished = onEffectFinished,
                        )
                        is LivesEffect.Gained -> BundledLottie(
                            name = "heart-refill",
                            tint = palette.secondary,
                            modifier = Modifier.size(glyph * 3f),
                            onFinished = onEffectFinished,
                        )
                    }
                }
            }
        }
    }
}

private val Int.dp get() = androidx.compose.ui.unit.Dp(this.toFloat())
